mp2-v1/tests/common/cases/table_source.rs

```diff
@@ -1436,3 +1460,416 @@ fn evm_word_column_info(table_info: &[ColumnInfo]) -> Vec<SlotEvmWordColumns> {
+                            // so we update 2 columns at random
+                            [0; 2].map(|_| {
+                                non_primary_key_columns
+                                    [rng.gen_range(0..non_primary_key_columns.len())]
+                                .clone()
+                            })
+                        } else {
+                            [
+                                non_primary_key_columns
+                                    [rng.gen_range(0..non_primary_key_columns.len())]
+                                .clone(),
+                                primary_key_columns[rng.gen_range(0..primary_key_columns.len())]
+                                    .clone(),
+                            ]
+                        };
+                        rows_to_update
+                            .into_iter()
+                            .zip(columns_to_update)
+                            .flat_map(|(row_index, column_data)| {
+                                self.update_non_indexed_column(row_index, &column_data)
+                                    .unwrap()
+                            })
+                            .collect()
+                    }
+                }
+            }
+            ChangeType::Silent => vec![], // no changes
+        }
+    }
+
+    pub(crate) fn generate_extraction_proof_inputs(
+        &self,
+        ctx: &mut TestContext,
+        proof_key: ProofKey,
+    ) -> Result<(ExtractionProofInput, HashOutput)> {
+        // convert current row values to row cells
+        let rows = self
+            .row_values
+            .iter()
+            .map(CellCollection::try_from)
+            .collect::<Result<Vec<_>>>()?;
+        let rng = &mut thread_rng();
+        // This could be computed from the table data according to any logic,
+        // here for simplicity we just generate it at random
+        let hash = HashOutput::from(array::from_fn(|_| rng.gen()));
+        // fetch previous hash from IVC proof for previous primary index, if any
+        let prev_hash = {
+            if let Ok(proof) = ctx.storage.get_proof_exact(&proof_key) {
+                // fetch previous hash from the proof
+                let pproof = ProofWithVK::deserialize(&proof)?;
+                let pis = PublicInputs::from_slice(&pproof.proof().public_inputs);
+                pis.block_hash_output()
+            } else {
+                // there is no previous primary index, so we generate previous hash at random
+                HashOutput::from(array::from_fn(|_| rng.gen()))
+            }
+        };
+
+        let metadata_hash = no_provable_metadata_hash(self.column_ids());
+        let input = ExtractionProofInput::Offchain(OffChainExtractionProof {
+            hash: OffChainRootOfTrust::Hash(hash),
+            prev_hash,
+            primary_index: self.last_update_epoch,
+            rows,
+            primary_key_columns: self.primary_key_column_ids(),
```

Comment:
This is confusing there should only be ONE primary key column no ?

Reply:
Not necessarily, a primary key may be composed of multiple columns: think of mapping of mappings tables, where the primary key is given by the keys of both the outer and the inner mappings. To clarify, primary key != primary index:: the primary key is the set of columns that uniquely identifies a row, while the primary index is just an index we build on a column (but it isn't necessarily unique for each row). I added a comment to `ColumnMetadata` of `OffchainTableArgs` data structure to clarify the meaning of primary key columns.

Reply:
Ahhh gotcha. ok. @delehef something we discussed this morning.